ksql-engine/src/test/java/io/confluent/ksql/codegen/CodeGenRunnerTest.java

```diff
@@ -528,4 +535,20 @@ public void testStringUDFExpr() throws Exception {
 
     }
 
+    @Test
+    public void testNestedUdf() throws Exception {
+        final Analysis analysis = analyzeQuery(
+            "SELECT CONCAT(SUBSTRING(col1,1,3),CONCAT('-',SUBSTRING(col1,4,5))) "
+            + "FROM codegen_test;");
+
+        ExpressionMetadata expressionEvaluator0 = codeGenRunner.buildCodeGenFromParseTree(
+            analysis.getSelectExpressions().get(0));
+
+        final Kudf[] udfs = expressionEvaluator0.getUdfs();
+
+        Object result0 = expressionEvaluator0.getExpressionEvaluator()
+                .evaluate(new Object[]{udfs[0], udfs[1], udfs[2], udfs[3], "Hello"});
+
+        assertThat(result0, equalTo("el-o"));
```

Comment:
For my understanding, what's happening is that when the same UDF is called twice in the same expression, both invocations share the same instance, and hence evaluate to the same value. So without this bugfix, `result0` would have been `el-el`?

Reply:
Without the bug fix the same instance is shared. How this affects the outcome depends on the UDF in question. If the UDF stores any state, e.g. `EXTRACTJSONFIELD` only parsers the JSON path on the first invocation, then the stored state is only valid for the first invocation. Subsequent invocations are executing with the wrong state.